### PR TITLE
[Example] Rgcn distributed training

### PR DESCRIPTION
**Pull request by classicsong**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/rgcn/experimental/README.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+In this example, we partition the ogbn-mag graph into 4 parts with Metis. The partitions are balanced with respect to
+the number of nodes, the number of edges and the number of labelled nodes.
+```bash
+python3 partition_graph.py --dataset ogbn-mag --num_parts 4 --balance_train --balance_edges
+```
+
+### Step 2: copy the partitioned data to the cluster
+DGL provides a script for copying partitioned data to the cluster. The command below copies partition data
+to the machines in the cluster. The configuration of the cluster is defined by `ip_config.txt`.
+The data is copied to `~/rgcn/ogbn-mag` on each of the remote machines. `--rel_data_path` specifies the location of the partitioned data (generated in Step 1). `--part_config`
```
Comment (zheng-da) [inline examples/pytorch/rgcn/experimental/README.md:21]: `--rel_data_path` specifies the relative path in the workspace where the partitioned data will be stored.

File: examples/pytorch/rgcn/experimental/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+--num_servers 1 \
+--num_samplers 4 \
+--part_config data/ogbn-mag.json \
+--ip_config ip_config.txt \
+"python3 entity_classify_dist.py --graph-name ogbn-mag --dataset ogbn-mag --fanout='25,25' --batch-size 512  --n-hidden 64 --lr 0.01 --eval-batch-size 16  --low-mem --dropout 0.5 --use-self-loop --n-bases 2 --n-epochs 3 --layer-norm --ip-config ip_config.txt  --num-workers 4 --num-servers 1 --sparse-embedding  --sparse-lr 0.06"
+```
+
+We can get the performance score at the second epoch:
+```
+Val Acc 0.4323, Test Acc 0.4255, time: 128.0379
```
Comment (zheng-da) [inline examples/pytorch/rgcn/experimental/README.md:53]: this is the performance of DGL's embedding or Pytorch's sparse embedding?

File: examples/pytorch/rgcn/experimental/entity_classify_dist.py [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+                self.node_embeds = dgl.distributed.DistEmbedding(g.number_of_nodes(),
+                                                                 self.embed_size,
+                                                                 embed_name,
+                                                                 init_emb)
+            else:
+                self.node_embeds = th.nn.Embedding(g.number_of_nodes(), self.embed_size, sparse=self.sparse_emb)
+                nn.init.uniform_(self.node_embeds.weight, -1.0, 1.0)
+        else:
+            self.node_embeds = th.nn.Embedding(g.number_of_nodes(), self.embed_size)
+            nn.init.uniform_(self.node_embeds.weight, -1.0, 1.0)
```
Comment (zheng-da) [inline examples/pytorch/rgcn/experimental/entity_classify_dist.py:164]: do we need to provide an option of using dense embedding?

File: examples/pytorch/rgcn/experimental/entity_classify_dist.py [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+        Returns
+        -------
+        tensor
+            embeddings as the input of the next layer
+        """
+        embeds = th.empty(node_ids.shape[0], self.embed_size)
+        for ntype in range(self.num_of_ntype):
+            assert features[ntype] is None, 'Currently Dist RGCN only support non input feature'
+            loc = node_tids == ntype
+            embeds[loc] = self.node_embeds(node_ids[loc])
```
Comment (zheng-da) [inline examples/pytorch/rgcn/experimental/entity_classify_dist.py:188]: this is differentiable? if the underlying is index_copy, this seems very slow.

File: examples/pytorch/rgcn/experimental/entity_classify_dist.py [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+        drop_last=False)
+
+    test_sampler = NeighborSampler(g, [-1] * args.n_layers, dgl.distributed.sample_neighbors)
+    # Create DataLoader for constructing blocks
+    test_dataloader = DistDataLoader(
+        dataset=test_nid.numpy(),
+        batch_size=args.batch_size,
+        collate_fn=test_sampler.sample_blocks,
+        shuffle=False,
+        drop_last=False)
```
Comment (zheng-da) [inline examples/pytorch/rgcn/experimental/entity_classify_dist.py:320]: getting all neighbors can be very slow.

File: tools/launch.py [hunk 6]
```diff
@@ -44,7 +44,10 @@ def submit_jobs(args, udf_command):
                 port = DEFAULT_PORT
                 hosts.append((ip, port))
             else:
-                raise RuntimeError("Format error of ip_config.")
+                ip = result[0]
+                port = int(result[1])
+                hosts.append((ip, port))
+                #raise RuntimeError("Format error of ip_config.")
```
Comment (zheng-da) [inline tools/launch.py:50]: why modify this?

File: examples/pytorch/rgcn/experimental/README.md (hunk 2 above)
Comment (classicsong) [reply examples/pytorch/rgcn/experimental/README.md:53]: Pytorch's sparse embedding
I give the command also.

File: examples/pytorch/rgcn/experimental/entity_classify_dist.py (hunk 3 above)
Comment (classicsong) [reply examples/pytorch/rgcn/experimental/entity_classify_dist.py:164]: Dense can also work in this dataset.
This is an example, we can show different choices.

File: examples/pytorch/rgcn/experimental/entity_classify_dist.py (hunk 4 above)
Comment (classicsong) [reply examples/pytorch/rgcn/experimental/entity_classify_dist.py:188]: When we introduce input features later, we have to do this.
This is not slow as the shape of loc and node_tids is not very large.

File: tools/launch.py (hunk 6 above)
Comment (classicsong) [reply tools/launch.py:50]: I forgot to revert it back

File: examples/pytorch/rgcn/experimental/entity_classify_dist.py (hunk 5 above)
Comment (classicsong) [reply examples/pytorch/rgcn/experimental/entity_classify_dist.py:320]: Testing suppose to use full neighbor sample.

File: examples/pytorch/rgcn/experimental/README.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
+## Distributed training
+
+This is an example of training RGCN node classification in a distributed fashion. Currently, the example only support training RGCN graphs with no input features. The current implementation follows ../rgcn/entity_claasify_mp.py.
+
+To train RGCN, it has four steps:
+
+### Step 1: partition the graph.
```
Comment (aksnzhy) [inline examples/pytorch/rgcn/experimental/README.md:7]: We add a step-0 in the graphsage example. Consider add it.

File: examples/pytorch/rgcn/experimental/README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+to the machines in the cluster. The configuration of the cluster is defined by `ip_config.txt`.
+The data is copied to `~/rgcn/ogbn-mag` on each of the remote machines.
+`--rel_data_path` specifies the relative path in the workspace where the partitioned data will be stored.
+`--part_config` specifies the location of the partitioned data in the local machine (a user only needs to specify
+the location of the partition configuration file). `--script_folder` specifies the location of the training scripts.
+```bash
+python ~/dgl/tools/copy_files.py --ip_config ip_config.txt \
+                                 --workspace ~/rgcn --rel_data_path data \
+				 --part_config data/ogbn-mag.json \
+			         --script_folder ./
```
Comment (aksnzhy) [inline examples/pytorch/rgcn/experimental/README.md:29]: Use ./ as the code folder can bring somr issues. If user download the their data in this folder. The data will be copied to all the machines.

File: examples/pytorch/rgcn/experimental/entity_classify_dist.py (hunk 5 above)
Comment (zheng-da) [reply examples/pytorch/rgcn/experimental/entity_classify_dist.py:320]: it's fine. the accuracy difference should be very small. at least the validation shouldn't use full neighbors.

Comment (classicsong) [reply examples/pytorch/rgcn/experimental/entity_classify_dist.py:320]: If we want to chose testing result based on validation set, I think validation should also use full neighbors.

Comment (BarclayII) [reply examples/pytorch/rgcn/experimental/entity_classify_dist.py:321]: In the current RGCN implementation we use neighbor sampling for validation because computing the validation accuracy is a reasonably frequent operation and including some randomness in neighbor aggregation for model selection is usually OK.  We do offline inference for testing where we take full neighborhood layer-by-layer however.

Comment (BarclayII) [reply examples/pytorch/rgcn/experimental/entity_classify_dist.py:321]: @classicsong said that there is little difference in speed with full-neighbor offline inference and testing with full neighbor DataLoader.
